### PR TITLE
[RAY AIR][DOC][TorchTrainer] Rewrote the TorchTrainer code snippet as a working example

### PR DESCRIPTION
**Pull request by dmatrix**
Signed-off-by: Jules Damji <jules@anyscale.com>
- Rewrote the code snippet as it was not working
- Removed python-code block directives; instead use testcode and testoutput. This will test code if it runs in the CI
- Ignore the output since we get loads of output from the three workers
- Assert that the loss converges with the training data within specified epochs
- Tested code end-to-end with this [script](https://github.com/dmatrix/ray-code-snippets/blob/main/rayair/train/torch/test_torch_trainer.py)

## Checks

- [x ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/train/torch/torch_trainer.py [hunk 1]
```diff
@@ -22,13 +22,14 @@ class TorchTrainer(DataParallelTrainer):
     The ``train_loop_per_worker`` function is expected to take in either 0 or 1
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:22]: In the paragraph above this, it says "already" twice in the sentence -- it would be great to also fix this :)

File: python/ray/train/torch/torch_trainer.py [hunk 2]
```diff
@@ -22,13 +22,14 @@ class TorchTrainer(DataParallelTrainer):
...
+    .. testcode::
 
         def train_loop_per_worker():
             ...
 
-    .. code-block:: python
+    .. testcode::
 
+        from typing import Dict
         def train_loop_per_worker(config: Dict):
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:33]: Ideally this would have a bit more typing like `Dict[str, Any]` (not sure what exactly the format here is) and also link to the format of the dict if possible :)

File: python/ray/train/torch/torch_trainer.py [hunk 3]
```diff
@@ -45,32 +46,33 @@ def train_loop_per_worker(config: Dict):
     Inside the ``train_loop_per_worker`` function, you can use any of the
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:46]: Ideally there would also be an example for the above paragraph somewhere, we can feel free to do that in another PR.

Comment (pcmoritz) [reply python/ray/train/torch/torch_trainer.py:46]: (You can discard this, I saw the usage is already shown in the example below -- maybe add `(see example below)`.

File: python/ray/train/torch/torch_trainer.py [hunk 4]
```diff
@@ -45,32 +46,33 @@ def train_loop_per_worker(config: Dict):
...
     :ref:`Ray AIR session methods <air-session-ref>`.
 
-    .. code-block:: python
+    .. testcode::
 
         def train_loop_per_worker():
             # Report intermediate results for callbacks or logging and
             # checkpoint data.
+            #
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:54]: I feel like it was better without this line but if you prefer feel free to keep it :)

File: python/ray/train/torch/torch_trainer.py [hunk 5]
```diff
@@ -45,32 +46,33 @@ def train_loop_per_worker(config: Dict):
...
             session.report(...)
 
-            # Returns dict of last saved checkpoint.
+            # Session returns dict of last saved checkpoint.
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:57]: Say `Get dict of last saved checkpoint.` here? `Session returns` is a little confusing I think :)

File: python/ray/train/torch/torch_trainer.py [hunk 6]
```diff
@@ -103,65 +106,106 @@ def train_loop_per_worker():
...
+            # Define your network structure
             class NeuralNetwork(nn.Module):
                 def __init__(self):
                     super(NeuralNetwork, self).__init__()
                     self.layer1 = nn.Linear(input_size, layer_size)
-                    self.relu = nn.ReLU()
                     self.layer2 = nn.Linear(layer_size, output_size)
 
                 def forward(self, input):
-                    return self.layer2(self.relu(self.layer1(input)))
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:120]: I would either keep the ReLU layer here or have only one linear layer -- composing two linear layers doesn't do anything and it would likely be confusing to users :)

File: python/ray/train/torch/torch_trainer.py [hunk 7]
```diff
@@ -103,65 +106,106 @@ def train_loop_per_worker():
...
-                        checkpoint=Checkpoint.from_dict(
-                            dict(epoch=epoch, model=model.state_dict()
-                        ),
+                        # Print what's happening with loss per 30 epochs
+                        if epoch % 30 == 0:
+                            print(f"epoch: {epoch}/{num_epochs}, loss: {loss:.3f}")
+
+                    # Report and record metrics, checkpoint model at end of each
+                    # epoch
+                    session.report({"loss": loss.item(), "epoch": epoch},
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:168]: This is confusing since `epoch` is both here and below, @amogkam can you recommend how to do this? Most users will follow the example, so we should make sure we do this well :)

File: python/ray/train/torch/torch_trainer.py [hunk 8]
```diff
@@ -103,65 +106,106 @@ def train_loop_per_worker():
...
+    .. testoutput::
+
+        ...
+
     Args:
+
         train_loop_per_worker: The training function to execute.
             This can either take in no arguments or a ``config`` dict.
-        train_loop_config: Configurations to pass into
+            train_loop_config: Configurations to pass into
```
Comment (pcmoritz) [inline python/ray/train/torch/torch_trainer.py:208]: The indentation should be kept here, right? Otherwise it won't render correctly :)

**Review by pcmoritz — APPROVED** [review]
Thanks for doing this, this is great! There are a few small comments you should address before merging :)

File: python/ray/train/torch/torch_trainer.py (hunk 1 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:22]: good catch. Fixed

File: python/ray/train/torch/torch_trainer.py (hunk 2 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:33]: Yeah, we can add some typing.

File: python/ray/train/torch/torch_trainer.py (hunk 4 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:54]: Since the code is incomplete `session.report(...)` and `session.get_checkpoint()`, nice to explain it with a comment

File: python/ray/train/torch/torch_trainer.py (hunk 5 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:57]: Yes, Get x makes sense then returns, since it's an explicit method call to sessin.get_xxx

File: python/ray/train/torch/torch_trainer.py (hunk 7 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:168]: One is reporting the loss per epoch as metrics, the other is there for checkpoint per epoch. Nice to have that metrics per epoch. If @amogkam feels strongly that we should not include "epoch" in the metrics to report, then I can remove that entity.

File: python/ray/train/torch/torch_trainer.py [hunk 9]
```diff
@@ -103,65 +106,106 @@ def train_loop_per_worker():
...
                 train_loop_per_worker=train_loop_per_worker,
                 scaling_config=scaling_config,
+                run_config=run_config,
                 datasets={"train": train_dataset})
+
             result = trainer.fit()
 
+            # Get the loss metric from TorchCheckpoint tuple data dictionary
+            best_checkpoint_loss = result.metrics['loss']
+            # print(f"best loss: {best_checkpoint_loss:.4f}")
```
[the remark this answers is not in the document]
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:194]: Yeah a bit redundant since the code is self-explanatory.

File: python/ray/train/torch/torch_trainer.py (hunk 8 above)
Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:208]: The Args name: should be indented on the same level. That is:

```
Args:
        arg_1:  ...
        arg_2: ...
```

Comment (dmatrix) [reply python/ray/train/torch/torch_trainer.py:207]: Screen Shot 2022-11-21 at 3 25 39 PM
They seem to render properlyl

**Comment by dmatrix** [comment]
@pcmoritz can we merge this and [30637](https://github.com/ray-project/ray/pull/30637) if you don't see any issues.

**Comment by pcmoritz** [comment]
Looks like some of the hugging face servers are down, which is independent of this PR, we can merge it after the tests ran.

**Comment by pcmoritz** [comment]
Signed-off-by: Jules Damji [jules@anyscale.com](mailto:jules@anyscale.com)

- Rewrote the code snippet as it was not working

- Removed python-code block directives; instead use testcode and testoutput. This will test code if it runs in the CI

- Ignore the output since we get loads of output from the three workers

- Assert that the loss converges with the training data within specified epochs

- Tested code end-to-end with this [script](https://github.com/dmatrix/ray-code-snippets/blob/main/rayair/train/torch/test_torch_trainer.py)
